js/privatebin.js

```diff
@@ -2123,6 +2101,42 @@ jQuery.PrivateBin = (function($, RawDeflate) {
             Helper.selectText($pasteUrl[0]);
         };
 
+        /**
+         * extracts URLs from given string
+         *
+         * if at least one is found, it disables the shortener button and
+         * replaces the paste URL
+         *
+         * @name   PasteStatus.extractUrl
+         * @function
+         * @param  {string} response
+         */
+        me.extractUrl = function(response)
+        {
+            if (typeof response === 'object') {
+                response = JSON.stringify(response);
+            }
+            if (typeof response === 'string' && response.length > 0) {
+                const shortUrlMatcher = /https?:\/\/[^\s]+/g;
+                const shortUrl = (response.match(shortUrlMatcher) || []).filter(function(a) {
+                    return URL.canParse(a);
+                }).sort(function(a, b) {
```

Comment:
Why that sorting? Want to extract the shortest or longest URL hmm? This is kinda prone to fail… (e.g. what happens if the terms linked at https://somedomain.com/terms are shorter than https://somedomain.com/aHTZg61?) Maybe just use the first?

Or this would need to be configurable in the backend, the admin likely knows whether the first or shortest URL is the correct one…

Reply:
My guess is as good as yours, as I didn't write the code. I assumed it would order the longest first, since it subtracts the lengths, making the sign negative if b is longer than a, so b would be less than a = ranked higher. So it finds the longest URL in the block of text, HTML or JSON given?

Reply:
But does that make sense given it should create a short link?

Again a reason to write it more descriptive, so having a function "getLongest" or "getShortest" URL?

Reply:
I really can't answer that without having more context. We need examples of what these API responses look like. Thats why I had asked for such samples to get shared in the ticket.

Lacking that, the test that I've added only generates a single random URL per test iteration and assumes the API response consists of a single URL. We got told that some APIs return some JSON message and others HTML, but I've not done much research - I really don't like the URL shortener feature, but as it is popular I'm willing to help support it. This logic was contributed by a kind member of the community and I don't want to remove it without knowing what the impact of doing so will be. I have to assume that the current logic works for people (except for the IDN bit we are trying to fix here) and was eager to do this small change because it lets us simplify one of those inscrutable regexes.

PS: Here is some doc I found from Yourls, which includes a sample JSON and XML response: https://yourls.org/readme.html#API;apireturn - in that example the longurl happens to be shorter than the shorturl, but I can see that with the above logic it will tend to find the paste URL itself, instead.

Reply:
Yeah ideally we should have had examples or a unit test when this was initially implemented, so we cannot break the behaviour for some setting.

But too late now I suppose, so feel free to merge.

Reply:
/cc @Haocen who apparently implemented this in https://github.com/PrivateBin/PrivateBin/commit/cd5ede5670985e6985d01d6cbce87b85f4c130da / https://github.com/PrivateBin/PrivateBin/pull/480

If you read this, some questions we have:
* Why is the longest URL filtered out?
* Can we have example responses that it works with?
* With which URL shorters has this been tested?

Reply:
Maybe it also was a bug and actually it was intended to get the short URL?
Because in PrivateBin the hash is sooo long that the long URL would almost always be the wrong one (unless you have a _huge_ domain name).
The question just arises, why was this issue never noticed?

Some ideas I have about this:
* Maybe include the YOURLS examples you have found in the tests. Or at least the structure (like JSON and ML but with a PB URL). This way we have a realistic test/use case. (It does not need to be randomized IMHO. We have a fixed target API in case of YOURLS, which we often mention in the project docs as a suggestion, so it's a good reference IMHO).
  Reason: In the future, we could implement a simple parsing trying some well-known URL shorters (like `shortURL in response`, if it is JSON etc.), maybe implemented as [a strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern), but IMHO this is future stuff, although it could be implemented right now, too. But I would like to keep it simple for now.)
* Maybe in the list of URLs retreived we could filter out our own URL i.e. the long URL? I.e. just filter it out `document.URL`?
   Reason: Then sorting does not matter that much and this is clearly the wrong URL.
   Note: I also thought just filtering out the current domain, but that is prone to fail, if someone chooses to host the URL shortener on the same domain as the PrivateBin instance itself.
* Otherwise I still personally feel like choosing the longest URL is a bug, but of course I cannot be certain.

Reply:
Hi, 👋
It's been a while.
The original logic is to find all URLs in the response and get the shortest one.
In a few URL shortener I tested, including TinyURL and YOURLS, they tend to include the original URL as well as the shortened URL.

Given that background, the logic for parsing an unknown structure response from URL shortener would be:
- extract Any URL like strings from the response.
- sort them from shortest to longest order.
- take the shortest one from all URLs.

It was working as expected when I implemented it and you can refer to how `sort` function works here:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

Reply:
Ok, let me merge this as is, but then I'll create another branch with additional tests based on the samples found for Yourls. Then we'll at least learn about the behaviour with short- and longurl parameters and can explore if this can be a bug or not and fix it if found necessary.